### ISSUE TITLE
Handle case-insensitive file systems in a better way

### ISSUE THREAD

**Issue by andrejbauer** [agda/agda#5719]
Where do I report bugs?

```agda
open import Relation.Binary.PropositionalEquality

module Error where

  data N : Set where
    Z : N
    S : N → N

  infixr 5 _+_

  _+_ : N → N → N
  m + Z = m
  m + S n = S (m + n)

  problem : ∀ a → S Z + a ≡ S a
  problem Z = refl
  problem (S a) = trans (problem (S a)) refl

-- An internal error has occurred. Please report this as a bug.
-- Location of the error: __IMPOSSIBLE__, called at src/full/Agda/TypeChecking/Errors.hs:296:18 in
-- Agd-2.6.2-7cfdab78:Agda.TypeChecking.Errors
```

System info:

```
% uname -a
Darwin OA 20.6.0 Darwin Kernel Version 20.6.0: Mon Aug 30 06:12:21 PDT 2021; root:xnu7195.141.6~3/RELEASE_X86_64 x86_64

% agda --version
Agda version 2.6.2
```

Do you need anything else?

**Comment by Trebor-Huang** [comment]
A random guess: since you are on a case-insensitive system, does your module name agree exactly with your file name (including case)? See #5508

**Comment by jespercockx** [comment]
I failed to reproduce the error on Linux, so it could be something platform-related indeed. Can someone on Apple try if they can reproduce this?

**Comment by andreasabel** [comment]
> A random guess: since you are on a case-insensitive system, does your module name agree exactly with your file name (including case)? See #5508

Yes, @andrejbauer :
- What is the filename of module `Error`?
- Does the issue go away with Agda-2.6.2.1?

**Comment by andreasabel** [comment]
The is likely no MWE coming.  Closing.

**Comment by andrejbauer** [comment]
Sorry for the delay. Using Agda 2.6.2 on MacOS, I can confirm that the problem is caused by case-insensitive file names. When I make sure that the file name matches, I get a reasonable error (about termination checking).

Is there a way for Agda to fail more graciously, or should we just blame Apple for this bug?

**Comment by gallais** [comment]
Looks like `Agda.Interaction.FindFile` already has the right logic. Should this ifdef:

https://github.com/agda/agda/blob/798be60d51a56a8c74cfd309c1498b070240e686/src/full/Agda/Utils/FileName.hs#L107-L114

also be activated for mac?

**Comment by andreasabel** [comment]
> Sorry for the delay. Using Agda 2.6.2 on MacOS, I can confirm that the problem is caused by case-insensitive file names. When I make sure that the file name matches, I get a reasonable error (about termination checking).
> 
> Is there a way for Agda to fail more graciously, or should we just blame Apple for this bug?

No, sorry @andrejbauer, this it totally Agda's fault.  I am on macOS myself, so I would be curious to reproduce the issue, but I was lacking information (i.e., the spelling of the filename of the module).  Issue #5508 could be related, which is fixed on Agda 2.6.2.1.  Could you please try to reproduce the issue on 2.6.2.1 (or complete the information so I can reproduce it)?

**Comment by andreasabel** [comment]
> Looks like `Agda.Interaction.FindFile` already has the right logic. Should this ifdef:
> 
> https://github.com/agda/agda/blob/798be60d51a56a8c74cfd309c1498b070240e686/src/full/Agda/Utils/FileName.hs#L107-L114
> 
> also be activated for mac?

Trying to remember my thought process here.  OK:
- For one, fixing `doesFileExistCaseSensitive` according to the _compile-time_ OS (rather than the _runtime_ OS) is a hack in the first place (think cross-compilation), but well.
- But then, case-sensitivity is not really an _OS_-specific thing, but a it is _file-system_ specific, if I am not mistaken. You can access a case-insensitive system also from Linux.
- Finally, the `find` API is different for Window and Unix (BSD/Linux).  For Windows, we can ask to find all files matching a pattern, like `dir/file.agda`, whereas in Unix we only get to ask for the contents of `dir/` search through them.  This can be costly for large directories, and I did not want to have this running on every call to `doesFileExistsCaseSensitive`.

It would be optimal if there was a function that would tell you the case-sensitivity of a location (like `dir/`).  In the testsuite, I have some code that infers the case-sensitivity by experiment.  Not a hack to be proud of, but at least it should be correct: https://github.com/agda/agda/blob/719ba6644bdeeccbe5f3e7eb5ed53921013cdab1/test/Fail/Tests.hs#L93-L106

**Comment by nad** [comment]
[Apparently](https://www.collabora.com/news-and-blog/blog/2020/08/27/using-the-linux-kernel-case-insensitive-feature-in-ext4/) some recent Linux kernel supports file systems for which `floß` and `FLOSS` refer to the same file.

What is the goal with Agda's support for case-insensitive file systems? One goal is presumably to avoid making users confused. Should we also aim to ensure maximum portability? In that case we should perhaps refuse having (say) both `A.agda` and `a.agda` on a case-sensitive file system. However, should we then also refuse having both `floß.agda` and `FLOSS.agda`? The rules of the file system I mentioned above supposedly depend on the version of Unicode that the file system uses.

**Comment by andreasabel** [comment]
> In that case we should perhaps refuse having (say) both `A.agda` and `a.agda` on a case-sensitive file system.

Yes!  I definitely support this.

> However, should we then also refuse having both `floß.agda` and `FLOSS.agda`?

That sounds too complicated.  
If you are using non-ASCII Unicode characters in your filenames, you have to be prepared for trouble.  

I'd stick to the common case of upper/lowercase confusions.

**Comment by nad** [comment]
> That sounds too complicated.

```
ghci> :set -XOverloadedStrings 
ghci> :set -package text-icu
package flags have changed, resetting and loading new packages...
ghci> import Data.Text.ICU
ghci> toCaseFold False "FLOSS" == toCaseFold False "floß"
True
```

**Comment by nad** [comment]
Perhaps one should first normalise:
```
ghci> "ö" == "ö"
False
ghci> toCaseFold False "ö" == toCaseFold False "ö"
False
ghci> normalize NFD "ö" == normalize NFD "ö"
True
```

**Comment by andreasabel** [comment]
This suggests that `text-icu` provides the standard.  I can change my mind, knowing a standard exists...

**Comment by andreasabel** [comment]
Ping @andrejbauer for a self-contained reproducer.

**Comment by andreasabel** [comment]
Ping @andrejbauer for a self-contained reproducer.
